cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -346,6 +417,11 @@ private predicate simpleInstructionLocalFlowStep(Instruction iFrom, Instruction
   //
   // Flow through the partial operand belongs in the taint-tracking libraries
   // for now.
+  // TODO: To capture flow from a partial definition of an object (i.e., a field write) to the object
+  // we add dataflow through partial chi operands, but only if the chi node is not the chi node for all
+  // aliased memory.
+  iTo.getAnOperand().(ChiPartialOperand).getDef() = iFrom and not iTo.isResultConflated()
```

Comment:
I worry that the new cases will lead to field conflation like I saw in https://github.com/Semmle/ql/pull/3123#issuecomment-609818877.

Given `struct S { int x, y; }`, this case adds flow from a write to `s.x` into `s` itself. The new case for chi -> load above will in principle add flow from `s` to a read of `s.y`. In practice, we're likely to be saved because the aliased SSA will usually not model `s.y` as a partial read from `s`. But I'm sure it will sometimes. What if there's a copy `S s2 = s;` followed by `s2.x = unrelated` to create a chi node and then `sink(s2.y)`? Or what if `s` is sent across a function call?

Reply:
I've actually noticed that since we moved to PostUpdateNodes on `ChiInstructions` we can actually have a _tiny_ amount of field flow without this addition. So I've removed this case in d56284f in the hopes that we can get this into 1.24.

Reply:
If what you need is flow out of definition-by-reference, then you can perhaps do for data flow what I did for taint in https://github.com/Semmle/ql/pull/3219. That PR splits the `ChiPartialOperand` case in two, and the first of those two cases might be suitable for data flow rather than just taint.

Reply:
Won't that have the same problems regarding field conflation in the following program?
```cpp
void f(S* s) {
  s->x = source();
}

void g(S* s) {
  f(s);
  // s = chi(total: Uninitialized[s], partial: source())
  sink(s->y)
}
```

Reply:
Does this conflation require flow through output params?

I'd hope that with the first rule of #3219 (not the second) the `s->x = source()` assignment does not cause flow into `*s` (the whole struct). There is a phi node after that assignment, right?

Reply:
You're right. It doesn't cause flow into the whole struct. The flow through output params was just to generate a `WriteSideEffectInstruction`.

It does seem like adding the first flow rule from #3219 gives us a few additional results in various library tests. I'll create a separate PR for that.

Reply:
I've opened https://github.com/Semmle/ql/pull/3220. AFAIC it's not related to this PR, but the new flows in `library-tests` looks like we want to have this anyway.